### PR TITLE
This commit fixes the JSON parser and execute_python methods so they work better with GPT-3.5

### PR DESCRIPTION
**Pull request by SuperYeti**
1. There was a bug in the JSON fixer in the parser, it was never making it to the code that tries to extract JSON if the AI returned text before the JSON packet.
2. GPT-3.5-Turbo keeps trying to execute_python_file with code. Added a execute_python_code method that saves the python code to a UUID named file, that is then passed to the execute_python_file method. Also changed the drive binding for docker that is launched to RW so that execute python code is able to write output files etc.
3. Updated the prompt to try and get GPT-3.5-Turbo to respond with just JSON more consistently.
4. The debug set statement in main.py got removed again. Re-added it so debugging can be enabled from launching the script via terminal.
5. execute_code.py was not using safe_join to ensure that files outside the workspace directory can't be executed. Updated.

### PR Quality Checklist
- [X] My pull request is atomic and focuses on a single change.
- [X] I have thouroughly tested my changes with multiple different prompts.
- [X] I have considered potential risks and mitigations for my changes.
- [X] I have documented my changes clearly and comprehensively.
- [X] I have not snuck in any "extra" small tweaks changes

### CODE REVIEW

File: scripts/json_utils.py [hunk 1]
```diff
@@ -122,6 +122,11 @@ def correct_json(json_str: str) -> str:
                 if cfg.debug_mode:
                     print('json loads error - add quotes', e)
                 error_message = str(e)
-        if balanced_str := balance_braces(json_str):
-            return balanced_str
+        try:
+            if balanced_str := balance_braces(json_str):
+                return balanced_str
+        except json.JSONDecodeError as e:
+            if cfg.debug:
```
Comment (lekapsy) [inline scripts/json_utils.py]: Typo: should be if cfg.debug_mode: ?

**Review by jaspythonis — DISMISSED** [review]

**Comment by hoblin** [comment]
BTW how do you run it? I'm running it in docker contained and always have this error once it tries to execute a code.
```
-=-=-=-=-=-=-= COMMAND AUTHORISED BY USER -=-=-=-=-=-=-=
Saving code to file for history. Filename:'761c301c-7c91-484d-9c57-5900bc54fc1c.py' in Workspace Folder:'auto_gpt_workspace'.
Code:
'import os
import errno

parent_dir = './promptmaster.pro/pm_frontend/__tests__/Landing'

if not os.path.exists(parent_dir):
    os.makedirs(parent_dir)
'
Executing file '761c301c-7c91-484d-9c57-5900bc54fc1c.py' in workspace 'auto_gpt_workspace'
SYSTEM:  Command execute_python_code returned: Error: Error while fetching server API version: ('Connection aborted.', FileNotFoundError(2, 'No such file or directory'))
```
image

**Comment by SuperYeti** [comment]
That error means that docker isn't running when I was running from my
machine, it may be because it's trying to launch a docker within a docker
container that doesn't have it installed.

Warren

**Comment by hoblin** [comment]
Then it doesn't seem usable in my case because I'm trying to create files and directories inside `./auto_gpt_workspace` dir.

**Comment by SuperYeti** [comment]
> Then it doesn't seem usable in my case because I'm trying to create files and directories inside `./auto_gpt_workspace` dir.

Can you provide a dockerfile for your environment and how you've setup, steps to reproduce etc? I'd be happy to take a look and see if anything jumps out at me.

File: scripts/data/prompt.txt [hunk 2]
```diff
@@ -60,4 +62,4 @@ RESPONSE FORMAT:
     }
 }
 
-Ensure the response can be parsed by Python json.loads
+Be absolutely sure the response can be parsed by Python json.loads.
```
Comment (Wladastic) [inline scripts/data/prompt.txt:65]: I feel like this sentence will only encurage it to explain itself haha

**Comment by hoblin** [comment]
NM on a docker problem, I'll think it through first. Here is another error:

```
ASSISTANT THOUGHTS:
REASONING:
CRITICISM:
Warning: Failed to parse AI output, attempting to fix.
 If you see this warning frequently, it's likely that your prompt is confusing the AI. Try changing it up slightly.
Failed to fix ai output, telling the AI.
NEXT ACTION:  COMMAND = Error: ARGUMENTS = string indices must be integers, not 'str'
SYSTEM:  Command Error: threw the following error: string indices must be integers, not 'str'
Warning: Failed to parse AI output, attempting to fix.
 If you see this warning frequently, it's likely that your prompt is confusing the AI. Try changing it up slightly.
Failed to fix ai output, telling the AI.
Error: Invalid JSON
 I apologize for the error. Here is the next command JSON. ``` { "thoughts": { "text": "The website did not provide any information on
 ```
 
 It loops over and over again. We probably should expect some apologies right after "human input" or some errors.

**Review by jaspythonis — DISMISSED** [review]

File: scripts/main.py [hunk 3]
```diff
@@ -288,6 +288,10 @@ def parse_arguments():
         print_to_console("Speak Mode: ", Fore.GREEN, "ENABLED")
         cfg.set_speak_mode(True)
 
+    if args.debug:
+        print_to_console("Debug Mode: ", Fore.GREEN, "ENABLED")
+        cfg.set_debug_mode(True)
+
```
Comment (nponeccop) [inline scripts/main.py]: This change is unrelated. Roll it back.

**Comment by SuperYeti** [comment]
>

**Comment by CodeKickIO** [comment]
Why was this closed? I am still running into the
`Command execute_python_file returned: Error: Invalid file type. Only .py files are allowed. `
issue that this PR attempts to fix.

**Comment by SuperYeti** [comment]
I closed it because of the number of PR's that were pushed before this pr
have significantly changed the entire layout of the project including file
locations and layouts and it was going to be a mess to merge. I'm just
reviewing the latest iteration of the project to see which of the
modifications I had proposed are still even necessary and will reassess
readding then.

Warren
